### PR TITLE
Kfs 1

### PR DESCRIPTION
**Pull request by eirys**
slt
lis bien, fais pas chier
bisou

### CODE REVIEW

**Review by JonathanDUFOUR — COMMENTED** [review]
Globally, everything looks good, except for some little details sometimes that I have commented on.
Great work!

File: src/drivers/serial.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+#include "core.h"
+
+namespace serial {
+
+/// The I/O port used by the serial communication driver.
+constexpr const u16 PORT = 0x3F8;
+
+/// Initializes the serial port driver.
+///
+/// This function should be called once before starting to use the serial port.
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/drivers/serial.h:24]: Who gives a f

File: src/drivers/serial.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    while ((core::inB(PORT + 5) & 0x20) == 0) {
+        core::pause();
+    }
+
+    // Write the byte.
+    core::outB(PORT, byte);
+}
+
+/// Writes a slice of bytes to the serial port.
+void write_str(i8 const *bytes, u32 count = 128) {
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/drivers/serial.h:61]: Not my work

File: src/drivers/serial.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    while (count && *bytes) {
+        write_byte(*bytes);
+        bytes += 1;
+        count -= 1;
+    }
+}
+
+void write_nbr(u32 value) {
+    i8 buf[32];
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/drivers/serial.h:70]: Not my work

File: src/drivers/vga.h [hunk 4]
```diff
@@ -6,96 +6,44 @@
...
-    union {
-        struct { uint8_t x, y; };
-        uint8_t elem[2];
-    };
+static constexpr const u16 CONTROL_PORT = 0x3D4;
+static constexpr const u16 DATA_PORT = 0x3D5;
 
-}; // class Cursor
+static constexpr const u32 WIDTH = 80;
+static constexpr const u32 HEIGHT = 25;
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/drivers/vga.h:21]: Indeed

File: src/drivers/vga.h [hunk 5]
```diff
@@ -6,96 +6,44 @@
...
+    Indigo      = 0x9,
+    Chartreuse  = 0xA,
+    Sky         = 0xB,
+    Cherry      = 0xC,
+    Barbie      = 0xD,
+    Daffodil    = 0xE,
+    Immaculate  = 0xF,
+
+    First       = Carbon,
+    Last        = Immaculate
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/drivers/vga.h:46]: Readability

File: src/drivers/vga.h [hunk 6]
```diff
@@ -104,94 +52,131 @@ class Char {
...
+    Char& operator=(Char&& other) = default;
+    Char& operator=(const Char& other) = default;
     ~Char() = default;
 
     inline
-    explicit operator uint8_t() const {
-        return inner;
+    explicit operator u8() const {
+        return m_inner;
+    }
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/drivers/vga.h:77]: Doc

File: src/drivers/vga.h [hunk 7]
```diff
@@ -6,96 +6,44 @@
...
 
-}; // class Cursor
+static constexpr const u32 WIDTH = 80;
+static constexpr const u32 HEIGHT = 25;
 
-/**
- * @brief The one cursor.
-*/
-static Cursor cursor;
+u8* const  BUFFER = (u8*)0xB8000;
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/drivers/vga.h:23]: No

File: src/drivers/vga.h [hunk 8]
```diff
@@ -104,94 +52,131 @@ class Char {
...
+        for (u32 x = 0; x < WIDTH; x++) {
+            BUFFER[2 * ((y - 1) * WIDTH + x)] = BUFFER[2 * (y * WIDTH + x)];
+            BUFFER[2 * ((y - 1) * WIDTH + x) + 1] = BUFFER[2 * (y * WIDTH + x) + 1];
         }
     }
-    for (uint32_t x = 0; x < VGA_WIDTH; x++) {
-        VGA_BUFFER[2 * ((VGA_HEIGHT - 1) * VGA_WIDTH + x)] = 0;
-        VGA_BUFFER[2 * ((VGA_HEIGHT - 1) * VGA_WIDTH + x) + 1] = 0;
+    for (u32 x = 0; x < WIDTH; x++) {
+        BUFFER[2 * ((HEIGHT - 1) * WIDTH + x)] = 0;
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/drivers/vga.h:135]: Indeed

File: src/main.cpp [hunk 9]
```diff
@@ -1,40 +1,89 @@
...
+#include "lib.h"
 
-#define KERNEL_NAME         "megamimOS"
-#define KERNEL_NAME_LEN     9
+/* -------------------------------------------- */
 
 /**
  * @todo :P)
 */
 struct MultibootInfo {
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/main.cpp:25]: No

File: src/main.cpp [hunk 10]
```diff
@@ -1,40 +1,89 @@
...
+    for (;;) {
+        ui::KeyEvent event;
+        ui::TranslateResult result = layout.translate(ps2::poll(), event);
 
-    // Testing lol
-    // ps2::poll();
-    // vga::scrollDown();
+        switch (result) {
+            case ui::TranslateResult::Print:    winManager << event.m_character; break;
+            case ui::TranslateResult::Invalid:  /* return */ /* _panic(winManager); */ break;
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/main.cpp:83]: Work for another time

File: src/ui/cursor.h [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+#pragma once
+
+#include "types.h"
+#include "core.h"
+
+namespace vga {
+
+enum class CursorPosResult: u8 {
+    Valid,
+
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/ui/cursor.h:22]: Who gives a f

File: src/ui/cursor.h [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+*/
+struct Cursor final {
+public:
+    /* ---------------------------------------- */
+    /*                   DATA                   */
+    /* ---------------------------------------- */
+
+    union {
+        struct { u8 m_x, m_y; };
+        u8 m_elem[2];
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/ui/cursor.h:41]: Ok

File: src/ui/cursor.h [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+    /* ---------------------------------------- */
+
+    inline
+    u8& operator[](const u32 index) {
+        return m_elem[index];
+    }
+
+    inline
+    const u8& operator[](const u32 index) const {
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/ui/cursor.h:65]: Indeed

File: src/ui/cursor.h [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+        m_y = newY;
+        return CursorPosResult::Valid;
+    }
+
+    /**
+     * @brief Sets the cursor to the given position.
+     */
+    inline
+    void setTo(const i8 x, const i8 y) {
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/ui/cursor.h:107]: Shouldn't happen

File: src/ui/cursor.h [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+        m_x = x;
+        m_y = y;
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/ui/cursor.h:109]: No
Simply think before doing

File: src/ui/history.h [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+    inline
+    void replaceLine(const i8* line, const u8 lineIndex) {
+        // u8 index = 0;
+        // while (line[index] && index < vga::WIDTH) {
+        //     m_data[lineIndex][index] = line[index];
+        //     index++;
+        // }
+        // m_data[lineIndex][index] = '\0';
+    }
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/ui/history.h:49]: No

File: src/ui/layout/key_event.h [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+namespace ui {
+
+enum class Action: u8 {
+    Released    = 0,
+    Pressed     = 1,
+    Repeat      = 2, // Not used
+
+    First       = Released,
+    Last        = Repeat
+};
```
[the remark this answers is not in the document]
Comment (eirys) [reply src/ui/layout/key_event.h:28]: Inintelligible

**Pull request merged**